### PR TITLE
Add VALUE_IN UDF to support multi-valued column group-by without extra values

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
1. Refactor all layers above Projection
2. Transform layer treat all columns as expressions
3. Re-design the TransformFunction interface to support dictionary in UDF
4. Add UDF support to both aggrgation columns and group-by columns
5. VALUE_IN UDF can act on multi-valued group-by columns and returns only the values inside the UDF

TODO: refactor selection to work on top of transform so that we can add UDF to selection in the future

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/GroupByAstNode.java [hunk 1]
```diff
@@ -36,10 +37,9 @@ public void updateBrokerRequest(BrokerRequest brokerRequest) {
         // List of expression contains columns as well as expressions to maintain ordering of group by columns.
         groupBy.addToExpressions(groupByColumnName);
       } else {
-        FunctionCallAstNode functionCallAstNode = (FunctionCallAstNode) astNode;
-
-        // Remove all white-space until we start compiling expressions on broker side.
-        groupBy.addToExpressions(functionCallAstNode.getExpression());
+        // Compile to standard expression string
+        String expression = ((FunctionCallAstNode) astNode).getExpression();
+        groupBy.addToExpressions(TransformExpressionTree.compileToExpressionTree(expression).toString());
```
Comment (kishoreg) [inline pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/GroupByAstNode.java]: why are we parsing it here. Is this data structure serializable? How will this backward compatible?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/BaseTransformFunction.java [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+          _stringValuesMV[i] = new String[numValues];
+          copyToString(doubleValues[i], _stringValuesMV[i], numValues);
+        }
+        return _stringValuesMV;
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+
+  private static void copyToInt(long[] source, int[] dest, int length) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/BaseTransformFunction.java]: move all this methods to some util class. Also add comments on why we cant use Arrays.copy methods

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/MultiplicationTransformFunction.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import com.linkedin.pinot.core.plan.DocIdSetPlanNode;
+import com.linkedin.pinot.core.query.exception.BadQueryRequestException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+
+public class MultiplicationTransformFunction extends BaseTransformFunction {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/MultiplicationTransformFunction.java]: why are all renames showing up as deletion and addition.

File: pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/GroupByAstNode.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/com/linkedin/pinot/pql/parsers/pql2/ast/GroupByAstNode.java]: Here we just check whether the expression string can be compiled, and put the standard equivalent expression string into the group-by expressions. This way we can avoid invalid query and ensure the "groupByColumns" in the broker response is standard. Will update the comments

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/MultiplicationTransformFunction.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/MultiplicationTransformFunction.java]: This is not renaming, the interface is redesigned so too many changes, and git identifies it as deletion and addition.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/BaseTransformFunction.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/BaseTransformFunction.java]: Done

File: pinot-server/src/main/java/com/linkedin/pinot/server/starter/ServerBuilder.java [hunk 4]
```diff
@@ -69,9 +70,19 @@ private void initMetrics() {
...
+    for (String transformFunctionClassName : _serverConf.getTransformFunctions()) {
+      Class<TransformFunction> transformFunctionClass;
+      try {
+        transformFunctionClass = (Class<TransformFunction>) Class.forName(transformFunctionClassName);
+      } catch (ClassNotFoundException e) {
+        throw new RuntimeException("Failed to find transform function class of name: " + transformFunctionClassName);
+      }
+      transformFunctionClasses.add(transformFunctionClass);
+    }
+    TransformFunctionFactory.init(transformFunctionClasses);
```
Comment (kishoreg) [inline pinot-server/src/main/java/com/linkedin/pinot/server/starter/ServerBuilder.java]: check if the transform function already exists

Comment (Jackie-Jiang) [reply pinot-server/src/main/java/com/linkedin/pinot/server/starter/ServerBuilder.java]: Done

**Pull request merged**